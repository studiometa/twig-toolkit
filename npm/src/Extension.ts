// eslint-disable-next-line import/no-extraneous-dependencies
import type { extend } from 'twig';
import {
  renderStyleAttribute,
  renderAttributes,
  renderClass,
} from './Helpers/Html.js';
import { HtmlElement } from './Tags/HtmlElement.js';
import { EndHtmlElement } from './Tags/EndHtmlElement';

/**
 * Extend Twig.
 *
 * @param {import('twig').Twig} instance
 */
export const extension: Parameters<typeof extend>[0] = (instance) => {
  const { escape } = instance.exports.filters;

  instance.exports.extendFunction('html_styles', (value) =>
    renderStyleAttribute(instance, value)
  );

  instance.exports.extendFunction('html_attributes', (value) =>
    renderAttributes(instance, value)
  );

  instance.exports.extendFunction('html_classes', (value) =>
    renderClass(instance, value)
  );

  instance.exports.extendTag(new EndHtmlElement(instance));
  instance.exports.extendTag(new HtmlElement(instance));
};
