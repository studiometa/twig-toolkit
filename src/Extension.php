<?php
/**
 * @link https://github.com/studiometa/twig-toolkit
 * @copyright Studio Meta
 * @license https://github.com/studiometa/twig-toolkit/blob/master/LICENSE
 */

namespace Studiometa\TwigToolkit;

use Studiometa\TwigToolkit\Helpers\Html;
use Studiometa\TwigToolkit\Helpers\Url;
use Studiometa\TwigToolkit\TokenParser\ElementTokenParser;
use Twig\Extension\AbstractExtension;
use Twig\TokenParser\TokenParserInterface;
use Twig\TwigFunction;

/**
 * Twig extension class.
 *
 * @author Studio Meta <agence@studiometa.fr>
 * @since 1.0.0
 */
class Extension extends AbstractExtension
{
    /**
     * Returns the token parser instances to add to the existing list.
     *
     * @return TokenParserInterface[]
     */
    public function getTokenParsers()
    {
        return [
            new ElementTokenParser(),
        ];
    }

    /**
     * Returns a list of functions to add to the existing list.
     *
     * @return TwigFunction[]
     */
    public function getFunctions()
    {
        return [
            new TwigFunction('html_classes', [Html::class, 'renderClass']),
            new TwigFunction('html_styles', [Html::class, 'renderStyleAttribute']),
            new TwigFunction(
                'html_attributes',
                [Html::class, 'renderAttributes'],
                ['needs_environment' => true, 'is_safe' => ['html']]
            ),
            new TwigFunction(
                'merge_html_attributes',
                [Html::class, 'mergeAttributes']
            ),
            new TwigFunction(
                'twig_toolkit_url',
                [Url::class, 'fromString'],
            ),
        ];
    }
}
