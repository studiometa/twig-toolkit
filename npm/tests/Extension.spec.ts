import { TwingEnvironment } from 'twing';
// eslint-disable-next-line import/no-unresolved
import { Extension, ExtensionName } from '../src/Extension.js';
