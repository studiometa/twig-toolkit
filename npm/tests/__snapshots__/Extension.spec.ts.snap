// Vitest Snapshot v1

exports[`The \`{{ html_attributes() }}\` Twig function > should encode as JSON complex attributes 1`] = `" data-options=\\"&#x7B;&quot;log&quot;&#x3A;true&#x7D;\\""`;

exports[`The \`{{ html_attributes() }}\` Twig function > should not render falsy attributes 1`] = `" autofocus checked"`;

exports[`The \`{{ html_attributes() }}\` Twig function > should render attributes 1`] = `" aria-hidden=\\"true\\" required class=\\"block foo\\" id=\\"foo\\""`;

exports[`The \`{{ html_styles() }}\` Twig function > should render inline CSS. 1`] = `"margin-top&#x3A;&#x20;10px&#x3B;&#x20;display&#x3A;&#x20;none&#x3B;"`;
